### DISCUSSION TITLE
Is it possible to specify maximum bond dimension when using CircuitToEinsum?

### DISCUSSION THREAD

**Discussion by GeorgePennington** [NVIDIA/cuQuantum#134] [Q&A]
I want to convert a `qiskit.QuantumCircuit` object to a Matrix Product State (more specifically the state prepared by the circuit) and specify either a) the maximum bond dimension in the MPS or b) the truncation threshold used for the SVD.

Is there a way to specify a) or b) when using `cuquantum.CircuitToEinsum`? I.e. can I modify `qiskit_basic.ipynb example` ([https://github.com/NVIDIA/cuQuantum/blob/main/python/samples/cutensornet/circuit_converter/qiskit_ba...](https://github.com/NVIDIA/cuQuantum/blob/main/python/samples/cutensornet/circuit_converter/qiskit_basic.ipynb)) to do this? 

I understand that I can contract the gates at a low level and specify an `algorithm` with specific `max_extent` or `rel_cutoff` (see [https://github.com/NVIDIA/cuQuantum/blob/main/python/samples/cutensornet/tn_algorithms/mps_algorithm...](https://github.com/NVIDIA/cuQuantum/blob/main/python/samples/cutensornet/tn_algorithms/mps_algorithms.ipynb)) but is the same thing possible at the high level using `CircuitToEinsum`?

**Comment by yangcal** [comment]
Hello George,

Thanks for the question. Note here that `CircuitToEinsum`, as the name suggests, it's a converter type class rather than a simulator. Currently the outputs from `CircuitToEinsum` can be parsed to `cuquantum.contract` to compute corresponding quantities and it does not have direct support for MPS simulation capability. 

In order to perform MPS simulation with bond or value truncation, you can parse the gates `CircuitToEinsum.gates` to the example MPS utility functions in `mps_algorithm.ipynb` notebook and add args to allow `max_extent` or `rel_cutoff` in the decomposition. 

Additionally, we have python bindings for our state APIs which can also compute the final MPS state with or without truncations. Please refer to the documentation or [examples](https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/examples/mps_amplitudes.html) for more details. We are also working on a more pythonic solution for python users so please stay tuned for our future releases. 

Thanks

**Reply by GeorgePennington** [reply]
Thanks for response and help!

Despite `CircuitToEinsum` being a converter class, it is also possible to call `CircuitToEinsum.reduced_density_matrix()` or `.amplitude()`. To clarify, does that mean that those methods are using the state vector APIs? Or are they using tensor network contractions without truncating? Thank you again in advance for your help!

**Reply by yangcal** [reply]
If you call `CircuitToEinsum.reduced_density_matrix()/amplitude()`, you'll get expression and the operands. These expression and operands can then be directly parsed to `cuquantum.contract` to compute the outcome using cuTensorNet. We don't have pythonic APIs that directly handle expression and operands.
